keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,741 @@
+
+- Systems such as real-time trading system or 5G CNFs (User Plain Function, UPF) employ DPDK to guarantee low-latency packet processing. DPDK (*Data Plane Development Kit*) consists of a set of libraries designed for packet processing acceleration managed in software, in the userspace. DPDK requires dedicated resources (such as exclusive CPUs, hugepages, and DPDK-compatible Network Interface Card), and most importantly, the alignment of resources to the same NUMA node so as to prevent an unexpected performance degradation due to inter-node (between NUMA nodes) communication overhead. This creates the reason to guarantee the reservation of memory (and hugepages) and other computing resources (e.g. CPU cores and NICs) from the same NUMA node, for DPDK-oriented containers. One of key KPIs in 5G deployment is Ultra-Reliable and Low Latency Communications (URLLC) which is mission-critical. This accounts for why this proposal is also critical.
+
+#### Story 2 : Databases
+
+- Databases (e.g., Oracle, PostgreSQL, and MySQL) require relatively extensive amount of memory and hugepages to access massive volumes of data relatively efficiently. In order to reduce the latency caused by cross-NUMA memory access and sharing, all resources (CPU cores, memory, hugepages, and I/O devices) should be aligned to the same single NUMA node. This will serve the stability and performance enhancement significantly.
+```
+<<[UNRESOLVED]>>
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TBD
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TBD
+
+<<[/UNRESOLVED]>>
+```
+
+## Design Details 
+
+### The Concept of Node Map and Memory Maps
+
+*Memory Manager* has an internal database, i.e. `Node Map` that incorporates `Memory Maps`. This database is used to keep track of the memory reserved for already deployed containers in Guaranteed QoS class. `Node Map` concept was endorsed to support multi-NUMA feature because prior to that, single-NUMA feature depended solely on `Memory Maps`. For that foregoing purpose, `Node Map` object remembers the static configuration of disjont NUMA-node groups (pre-configured via `--multi-numa-groups` *kubelet* flag) in `Node` objects. Notice that in fact `Memory Map` offers the counters for tracking the memory; therefore, the map should not be understood as a map that allows for reservation of ranges of memory or consecutive blocks of memory. 
```

Comment:
I removed the statement:
> " Node Map concept was endorsed to support multi-NUMA feature because prior to that, single-NUMA feature depended solely on Memory Maps"

because I added this one:
> Technically, the *single-NUMA* strategy is a special case of the *multi-NUMA* strategy, and thus, no separate implementation was developed for them. Still, the proposal differentiates between the *single-NUMA* and the *multi-NUMA* strategies. First of all, it allows for a number of important use cases to be discussed and contrasted. Also, the *multi-NUMA* strategy is subjected to the additional configuration options that affect the way it performs.          


I will keep:
> `Node Map` object remembers the static configuration of disjont NUMA-node groups (pre-configured via `--numa-groups` *kubelet* flag) in `Node` objects.

until we describe "dynamic" NUMA node grouping in this new section:


> <<[UNRESOLVED]>>
> 
> \#\#\# Static versus dynamic grouping of NUMA nodes, and why `--numa-groups` flag was proposed
> 
> In the section above, we explained why NUMA groups must be disjoint.
> The disjont set of NUMA groups can be configured either:
> * statically - it is why we introduced `--numa-groups` kubelet flag
> * dynamically - TBD... 
> 
> TBD: a dynamic approach is something that may lead to the eliminate of `--numa-groups` flag, and thus, the discussion about setting the default value for `--numa-groups` might be also not necessary anymore...
> 
> TBD: present cons & pros for both static and dynamic NUMA nodes grouping/assignment/allocation/etc.
> 
> <<[/UNRESOLVED]>>

@cynepco3hahue Could we discuss the "dynamic" approach on Slack?

I added the section above, and this another section:
> \#\#\# How to enable the guaranteed memory allocation over many NUMA nodes?

as you advised, in order to explain why:
1) the NUMA node groups must be disjoint (so that the guaranteed memory allocation can be ensured)
2) `--numa-groups` flag is needed